Add macOS support

eikehein opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This adds support for macOS to TR1X, which involved making the following changes:

- The OpenGL 2.1 support on macOS does not support GL_ARB_explicit_attrib_location and GL_EXT_gpu_shader4. I therefore added a code path to request an OpenGL 3.3 Core Profile context and appropriate shaders, as the additions of these extensions are part of 3.3 Core. This was done using unified shader files + also unifies the vertex and fragment shaders into single files, by providing preprocessor defines to the shader files. This approach is similar to OpenLara's and was requested by @rr-.

- A few changes to the frame timing code to use higher-precision types and not underflow in the calculation, due to the very high return value of SDL_GetPerformanceFrequency() on Apple Silicon. @rr- has already introduced similar changes in the 60fps branch, so the general approach should be aligned.

- Making a few 8-byte alignment fixes to struct definitions on struct layout. This is required for 64 bit alignment rules on ARM, but macOS' llvm toolchain also generates warnings on x86-64 otherwise.

- Extending the meson build system with macOS-specific logic that enables dynamic linking (macOS applications always dynamically link at least the System framework) and produces a macOS-style .app bundle on installation. Creating the .app bundle involves copying over dynamically linked libs and rewriting their link paths, which is performed by a new utility script `tools/mac/bundle_mac_dylibs.sh` - and yes, that's really how it's commonly done on macOS.

- A version of the TR1X app icon in macOS' .icns format, used by the app bundle and the installer image.

- An expanded README.

- Adding a GitHub workflow that builds TR1X and its dependencies for both arm64 and x86-64, creates a universal binary app bundle, as well as a macOS-style .dmg installer image (using the new `tools/mac/create_mac_installer.sh`). 

  The binaries and the installer image get code-signed and notarized, which is needed to distribute software on latest macOS. The code-signing and notarization steps require an Apple developer account and certificate, which I'm happy to sponsor for the project and which @rr- has agreed to. To that end the workflow file references a few secrets (cert, account id, etc.) that need to be set in the GitHub repo settings which I will make available.

  Should I get hit by a bus or something, the project could either simply stop signing/notarizing or get another Apple developer account and swap out the credentials, so there should be no fundamental bus number concerns.

  The final installer image comes in at 17.7 MB, which is not bad for two architectures vs. the x86-only 12.1 MB Windows installer.

eikehein committed:
specific: Fix frame timing on macOS

- `SDL_GetPerformanceFrequency` returns a very high value on Apple
  Silicon and requires an appropriate data type to store.
- The calculation of `delay_sec` could underflow if we're running
  behind and `target_counter - m_Counter` is negative, leading to
  a very high positive value and essentially freezing the
  rendering.

eikehein committed:
gfx: Add a OpenGL 3.3 core profile code path, unified shader files

- Vertex and fragment shaders are unified into single files.
- Add shader source pre-processing function that selects between shader
  type and OpenGL version by programmatically supplying "#define".
- Request a OpenGL 3.3 core profile context on macOS, as OpenGL 2.1
  contexts on the platform do not support GL_ARB_explicit_attrib_location
  or GL_EXT_gpu_shader4.
- Add an OpenGL 3.3 core profile codepath to all shaders.

eikehein committed:
types: Remove pragma forcing 1-byte alignment

arm/arm64 prefer 4- and 8-byte alignment, respectively; this lets
the compiler make up its own mind.

eikehein committed:
readme: Add macOS section

eikehein committed:
build: Add GitHub workflow for macOS

- Installs build basic dependencies via Macports.
- Builds universal arm64/x86-64 binaries of ffmpeg manually.
- Builds TR1X as a universal arm64/x86-64 executable.
- Uses `create_mac_installer` (Python) to create an installer .dmg disk image.
- Code-signs the binaries, app bundle and .dmg installer image.
- Notarizes the .dmg installer image.

eikehein commented:
For early review by @walkawayy ad @lahm86.

You can see an example run of the GitHub workflow here:

https://github.com/eikehein/TR1X/actions/runs/8126372817

The .dmg artifact is also downloadable there. @walkawayy has already tested this a bit on their Intel Mac (and I've tested it on a M1 Mac, both native and x86-64 via the Rosetta binary emulator).

Lint run is green: https://github.com/eikehein/TR1X/actions/runs/8126370992

eikehein commented:
Note that to fully merge/integrate this, we'll have to decide how to integrate the macOS workflow with `.github/release.yml`, i.e. how things call each other and how the artifact handover should work, but it shouldn't be too hard.

lahm86 commented on src/gfx/gl/program.c:
```diff
@@ -36,6 +37,50 @@ void GFX_GL_Program_Bind(GFX_GL_Program *program)
+    strcpy(processed_content, version_ogl33c);
+    strcat(processed_content, define_ogl33c);
+#else
+    stcpy(processed_content, version_ogl21);
```
Typo - `strcpy`

eikehein replied:
Thanks, fixed!

lahm86 commented on src/specific/s_clock.c:
```diff
@@ -32,8 +32,7 @@ int32_t S_Clock_SyncTicks(void)
             (double)(m_Counter - last_counter) / m_Frequency;
-        const double delay_sec =
-            (double)(target_counter - m_Counter) / m_Frequency;
+        const double delay_sec = abs(target_counter - m_Counter) / m_Frequency;
```
This generates the following warning.
```
warning: taking the absolute value of unsigned type ‘long long unsigned int’ has no effect [-Wabsolute-value]
```

eikehein replied:
Thanks, fixed in the proper way.

The underlying issue here is that on macOS, one of the very early frames ends up running behind, i.e. `target_counter - m_Counter` ends up being negative. The division by `m_Frequency` then causes an underflow, so `delay_sec` ends up with a large positive value. Multiplied by `* 1000`, and rendering essentially freezes as the next frame gets scheduled far in the future.

The idea for now is to simply not introduce any delay if we're already running behind, which I think is the best way to go with the current simple frame timing code.

In the future we probably want to implement some proper frame pacing and skip frames if we're running late, but this is best left to wait on @rr-'s interpolation work.

lahm86 reviewed (changes requested):
A couple of code comments below. Fixing the typo yields a build, but it fails to launch for me as follows.

![image](https://github.com/LostArtefacts/TR1X/assets/33758420/a6e2f674-9331-45fd-bdd1-1622aa9c5268)

[TR1X.log](https://github.com/LostArtefacts/TR1X/files/14486794/TR1X.log)

eikehein commented:
@lahm86 I think I spotted the shader issue. Give it another spin? :-) If this still doesn't work I'll do a Windows build to check over here.

I've also addressed the other comments.

Edit: One more push coming up to make the code formatting linter happy.

lahm86 reviewed (commented):
Plays now on Windows 👍 
Unfortunately, I don't own a Mac so can't test that, but I'm happy to go with the testing approach already taken.

I wonder regarding the shader changes if #795 will have been affected? Perhaps we should make a test build available once the PR has gone through further reviews for others to try as well.

eikehein commented:
> Plays now on Windows 👍 Unfortunately, I don't own a Mac so can't test that, but I'm happy to go with the testing approach already taken.

In addition to @walkawayy and myself, there's also Frjttr on the TEN Discord who's been testing it with Sabatu’s TR1 as a Customs testcase for me :). They're also using a M1/arm64 Mac.

> I wonder regarding the shader changes if #795 will have been affected? Perhaps we should make a test build available once the PR has gone through further reviews for others to try as well.

It shouldn't :). Basically, at the moment this preserves the GLSL 1.20 shaders in all situations except on __APPLE__, where it uses OpenGL 3.3 Core w/ GLSL 1.30 shaders instead as the OGL21/GLSL120 support on macOS isn't up to our extension requirements.

walkawayy reviewed (approved):
I don't know anything about the shader files honestly, but the changes look good to me. Don't notice any issues on my Intel Macbook Pro 2019 or Windows 11. Awesome job! I'll tag rr for review for when he has time.

rr- commented on .github/workflows/build_macos.yml:
```diff
@@ -0,0 +1,219 @@
+        with:
+          name: TR1X-Installer.dmg
+          path: ${{ github.workspace }}/installer/TR1X-Installer.dmg
+          compression-level: 0 # .dmg is already compressed.
```
Hi, it's best if we integrated this with the release job.
However, I'd want us to follow TR2X modernized and more modularized build process: https://github.com/LostArtefacts/TR2X/tree/develop/.github/ so it's fine to do this in a follow-up PR.

eikehein replied:
My suggestion is that after merging, you do the integration with the release workflow in a follow-up commit since you have the vision on that one :)

rr- replied:
Thanks, I'll do that, but I don't have a clear idea when this happens.

rr- commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 - added stack traces to logs for better crash debugging (#1165)
 - changed the way music timestamps are internally handled – resets music position in existing saves
+- added support for arm64 and Apple Silicon
+- added support for macOS builds
```
Let's put the `added` entries above `changed`
Let's also add a changelog entry for the changed shaders structure and compatibility as I believe the proposed changes should already help with certain hardware.

eikehein replied:
Done

rr- commented on meson.build:
```diff
@@ -323,4 +327,14 @@ executable(
+  install_subdir('bin/data', install_dir : 'Contents/Resources')
+  install_subdir('bin/shaders', install_dir : 'Contents/Resources')
+  install_data('tools/resources/icon.icns', install_dir : 'Contents/Resources')
+  install_data('tools/resources/Info.plist', install_dir : 'Contents')
```
Can we move these last two files into `tools/resources/mac/`, or `tools/mac/resources/` (whichever makes more sense to you)?

rr- commented on src/gfx/context.c:
```diff
@@ -118,6 +118,17 @@ void GFX_Context_Attach(void *window_handle)
+    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 3);
+    SDL_GL_SetAttribute(
+        SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
+#endif
```
Is it possible for us to convert this `#ifdef` into an abstract function that in `src/gfx/context_mac.c` contains these lines and in `src/gfx/context_generic.c` contains a no-op implementation, and then use conditional compilation in `meson.build` to choose between the implementations?

rr- commented on src/gfx/gl/program.c:
```diff
@@ -36,6 +37,50 @@ void GFX_GL_Program_Bind(GFX_GL_Program *program)
+    strcat(processed_content, define_ogl33c);
+#else
+    strcpy(processed_content, version_ogl21);
+#endif
```
Ditto – I'm OK with duplicating some of the code between mac-specific and generic implementation.
Alternatively, we can create `platform.h` `platform_mac.c` `platform_linux.c` that returns an enum `PLATFORM_MAC` `PLATFORM_LINUX` respectively etc., which we then can use it for things that only need this information at runtime and not compile time.

All in all, this is to try to stay away from `#ifdefs` as much as possible.

eikehein replied:
In the updated MR I'm using Meson's `configure_file` mechanism to generate a gfx_options.h that provides a define for the default backend, and sets it differently depending on the build platform.

The code is now ifdef-free, and the shader pre-processing function now takes a GFX_GL_BACKEND argument from an extensible enum, allowing us to do further work on e.g. GLES later on. This also means the code is now free of the platform notion again; instead, the build system opts macOS into preferring OpenGL 3.3.

rr- commented on src/global/types.h:
```diff
@@ -1933,14 +1941,6 @@ typedef struct REQUEST_INFO {
-    TEXTSTRING *background;
-    TEXTSTRING *moreup;
-    TEXTSTRING *moredown;
-    TEXTSTRING *texts[MAX_REQLINES];
```
Is there a reason why we needed to reorder the members?

eikehein replied:
Basically, on llvm/arm64, you want pointer members aligned to multiples of the 8-byte boundary. There's 3 ways to do this:

- Adding padding members, imho icky
- Doing weird tetris like having members that add up to multiples of 8 inbetween
- Putting all the pointers first

This seems the least icky?

rr- replied:
Sounds good, but what happens if they're not aligned? I think we have more structures like this BTW.

rr- replied:
All in all this seems to be pretty easy condition to break. I wonder if removing the `#pragma push(pack, 1)` from the beginning that forces the alignment would help? It's there mostly for legacy reason from the decompilation era. We should be good to remove it altogether for the most part, but I'm almost certain there are a couple of places left that use unsafe casts and rely on the member sizes/order. IMO it's better to track these and let the OS decide on data alignment for itself rather than fiddle with it by ourselves. Please LMK your thoughts.

eikehein replied:
Yep it's a good point, I'll investigate what this macOS toolchain does if I remove the fence and the attrib, maybe it just ends up being happy by itself

rr- commented on src/global/types.h:
```diff
@@ -1922,7 +1922,15 @@ typedef struct BOX_INFO {
 } BOX_INFO;
 
-typedef struct REQUEST_INFO {
+typedef struct __attribute__((aligned(8))) REQUEST_INFO {
```
Isn't this accomplished by `#pragma pack(push, 1)` code fence that should span this struct as well?

eikehein replied:
`pack(push, 1)` aligns to 1-byte boundaries, that attrib aligns to 8-byte boundaries.

rr- commented on tools/mac/bundle_mac_dylibs.sh:
```diff
@@ -0,0 +1,119 @@
+    update_links "$APP_BINARY_PATH"
+fi
+
+echo "Libraries for $APP_NAME copied and updated."
```
For more complex scripts, we try to stay away from Bash and Shell – can we convert this into a Python script using `argparse`?

rr- commented on src/specific/s_clock.c:
```diff
@@ -6,11 +6,11 @@
 
 static Uint64 m_Counter = 0;
-static double m_Frequency = 0.0;
+static Uint64 m_Frequency = 0.0;
```
I think it should be `0` or `0ull` or similar.

eikehein replied:
Whoops, forgot this one - updating, sec

rr- commented on tools/mac/create_mac_installer.sh:
```diff
@@ -0,0 +1,41 @@
+  --hide-extension "${APP_NAME}.app" \
+  --app-drop-link 600 185 \
+  "${DMG_NAME}" \
+  "${APP_BUNDLE_PATH}"
```
Likewise, let's try to use Python

rr- reviewed (commented):
Thanks. Awesome work. I left a couple of comments to let the code be more in line of what we are using or trying to follow – otherwise looks great!

eikehein committed:
build: Add support for macOS (app bundle, .dmg image) (#997)

- Adds a section for `darwin` (macOS) to meson.build that assembles
  an .app bundle folder structure in /tmp/TR1X.app.
- Adds the required `Info.plist` (app bundle metadata) and `icon.cns`
  (app icon in macOS .icns format, derived from the other icon in the
  codebase) files.
- The macOS build uses dynamic linking by default.
  `bundle_dylibs` (Python) uses `otool` to gather the libs from the
  linked executable and copies them into the app bundle, then
  rewrites the linkage paths in the executable. The dynamic
  libraries will typically come from e.g. Homebrew or MacPorts. This
  is a common way to do it.
- `create_mac_installer` (Python) uses the `create-dmg` program (also
  available from e.g. Homebrew) to bake the app bundle into a
  .dmg installer image. This will later be used by a GitHub CI/CD
  workflow together with code-signing steps.

Resolves #997.

rr- commented on bin/shaders/2d.glsl:
```diff
@@ -0,0 +1,36 @@
+void main(void) {
+    OUTCOLOR = TEXTURE(tex0, vertTexCoords);
+}
+#endif // VERTEX
```
Can we indent preprocessed fragments here and in the other shaders?
Eg

```
#ifdef VERTEX
    yadda yadda
#else 
    blabla boo boo
#endif
```

eikehein replied:
Will do

rr- commented on src/gfx/common.h:
```diff
@@ -12,3 +14,8 @@ typedef enum GFX_RENDER_MODE {
+typedef enum GFX_GL_BACKEND {
+    GFX_GL_21,
+    GFX_GL_33C,
+} GFX_GL_BACKEND;
```
Missing \n at EOF

eikehein replied:
Will do

rr- commented on src/gfx/gl/program.c:
```diff
@@ -36,6 +37,50 @@ void GFX_GL_Program_Bind(GFX_GL_Program *program)
+{
+    const char *version_ogl21 =
+        "#version 120\n#extension GL_ARB_explicit_attrib_location: "
+        "enable\n#extension GL_EXT_gpu_shader4: enable\n";
```
Can we make it like so:
```
const char *version_ogl21 =
    "blabla\n"
    "yadda yadda yadda\n"
    "bar qux\n";
```
so that `\n` always appears at the end of the string fragment?

eikehein replied:
Will do

rr- commented on src/gfx/gl/program.c:
```diff
@@ -50,7 +95,14 @@ void GFX_GL_Program_AttachShader(
+        Shell_ExitSystemFmt("Failed to pre-process shader source:  %s", path);
+    }
+
+    glShaderSource(shader_id, 1, (const char **)&processed_content, NULL);
```
Is this cast necessary?

rr- commented on bin/shaders/2d.glsl:
```diff
@@ -0,0 +1,38 @@
+void main(void) {
+    OUTCOLOR = TEXTURE(tex0, vertTexCoords);
+}
+#endif // VERTEX
```
Sorry to be repetitive but we're missing an \n here as well as the other shaders too. Whole we're at it let's make sure the other files have it as well.

rr- commented on src/global/types.h:
```diff
@@ -2036,8 +2034,6 @@ typedef struct SAMPLE_INFO {
     int16_t flags;
 } SAMPLE_INFO;
 
```
I'll confirm later today if the game code uses unsafe casts anywhere that night break this.

rr- replied:
I have checked all casts across the project and it seems the last unsafe operations were removed with the introduction of the XYZ_32 `struct`. I used this command to look for the casts:

```
grep -P '(?<!sizeof ?)\([^\(\),]*\*\)' **/*.c|grep -v -P '\((const )?(void|u?int(8|16|32)_t|char|bool|float|double) ?(\* ?)?(const ?)?\*\)'
```

eikehein commented:
Squashed / linearized history
